java/arcs/core/host/ParticleContext.kt

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.core.host
+
+import arcs.core.data.Plan
+import arcs.core.entity.Handle
+import arcs.core.host.api.Particle
+import arcs.core.storage.StorageProxy.StorageEvent
+import arcs.core.util.TaggedLog
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+/** Maximum number of times a particle may fail to be started before giving up. */
+const val MAX_CONSECUTIVE_FAILURES = 5
+
+/**
+ * Holds per-[Particle] context state needed by [ArcHost] to implement [Particle] lifecycle.
+ *
+ * @property particle currently instantiated [Particle] class
+ * @property planParticle the [Plan.Particle] used to instantiate [particle]
+ * @property dispatcher used when invoking the particle lifecycle methods
+ * @property particleState the current state the particle lifecycle is in
+ * @property consecutiveFailureCount how many times this particle failed to start in a row; used
+ *           to detect infinite-crash loop particles
+ */
+class ParticleContext(
+    val particle: Particle,
+    val planParticle: Plan.Particle,
+    val dispatcher: CoroutineDispatcher = Dispatchers.Unconfined,
```

Comment:
I think it can probably be Dispatchers.Default. There was an issue with Default not working with compilation to WASM or JS I think at one point, but we shouldn't let this block us using the right dispatcher.

Reply:
In practice this dispatcher is supposed to be a Scheduler.Dispatcher. `Default` would not be what we'd want.

Reply:
Changed this to accept a `Scheduler` (with null where not needed).